### PR TITLE
Use in memory cache for Kubernetes API discovery and HTTP requests

### PR DESCRIPTION
**Pull request by billiford**
*Note: I am currently just looking for feedback on implementation and features while this is being tested in our sandbox*.

This PR converts Go Clouddriver to use an in-memory cache store *by default* for both its Kubernetes API discovery and HTTP response caching. It is configurable by setting the environment variable `KUBERNETES_USE_DISK_CACHE` to `true`.

I rewrote the [memory/memcache.go](https://github.com/kubernetes/client-go/blob/master/discovery/cached/memory/memcache.go) to be logically similar to [disk/cached_discovery.go](https://github.com/kubernetes/client-go/blob/master/discovery/cached/disk/cached_discovery.go).

I added a roundtripper that uses `github.com/gregjones/httpcache` for its in-memory HTTP request caching. The disk caching implementation uses a directory to store these cached requests (https://github.com/kubernetes/client-go/blob/master/discovery/cached/disk/round_tripper.go).

Additionally we are now caching Kubernetes rest configs and their corresponding in-memory cache discovery clients (Cache All The Stuff!). The purpose of this is because generating these discovery clients is dependent on some config. Parts of the config can change (for example the `CAData` or `BearerToken`), so it's necessary to hold this information in memory and check if it's changed to know if we need to generate a new memory store for API discovery. Currently, we only check if the `CAData` or `BearerToken` have changed and then update the clients as needed.

I decided to put the mutex locks at the beginning of any function that makes live calls for discovery in the `memcache.go` file. This is very helpful since Go Clouddriver is often making concurrent calls to the cluster, which can result in multiple API discovery transactions taking place at once. Since we're no longer dealing with the overhead of writing and reading files this seemed logical and has the added benefit of a client only making a single API discovery transaction regardless of how many concurrent requests are made to the cluster.

Feedback or comments are welcome. Specifically:
- Should I add tests for the `internal/kubernetes/controller.go` file now that there's more logic in there?
- How do we feel about the name for the environment variable to use disk cache?
- Can we think of a better way to handle storing configs and caching clients in memory?

### CODE REVIEW

**Comment by billiford** [comment]
I'm still having an issue where new CRDs are not triggering API discovery.
image

**Comment by dmrogers7** [comment]
1. This is probably an argument for controller tests 
   image
2.  I'm okay with the env var name.  I suppose naming it KUBERNETES_CACHE_TYPE with supported values of `disk` and `memory` might future proof it if another type is ever introduced, but we can cross the bridge if it comes.
3. I would have `ourEntries` hold everything as the cached bytes and cast it to correct type in the functions
    ```golang
    type memCacheClient struct {
 	  delegate discovery.DiscoveryInterface

 	  // ttl is how long the cache should be considered valid
 	  ttl time.Duration

 	  mutex           sync.Mutex
 	  ourEntries      map[string][]byte
 	  ourTTLs         map[string]time.Time
 	  invalidated     bool
 	  // fresh is true if the cache is not populated or all requested entries
 	  fresh bool
    }
    ```
    Then "servergroups" is just another Entry, so there's no need for the `getCachedServerGroups` and `createCachedServerGroups` functions - just use the `getCachedEntry` and `createCachedEntry` functions.  This would mirror what is done in `disk/cached_discovery`.

**Comment by billiford** [comment]
1. Tests have been added.
2. For sure. If you want to request changes feel free too, but I think crossing that bridge later is fine.
3. Done.

This ended up being kind of a pain, but I got it working. It took a lot of understanding what the disk cache was doing. If you check the comments for the fields on the `memCacheClient` struct and compare the solution to the disk cache solution you should get an idea of what I'm doing here.

In short, the disk cache relies on files on the file system. Creating a new disk cache will not remove these files. I had to create two variables to hold
1. `entries`: think of these as files on a file system that will be maintained by caching the client itself.
2. `ourEntries`: just like the `ourFiles` field in the disk cache, this needs to be reset every time we grab the client from cache. Its purpose is to know if our `entries` are "fresh" (i.e. were they created during the same process that generated the mem cache client).

I still want to leave this sitting in sandbox for a bit.

File: internal/kubernetes/cached/memory/memcache.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+	return rs, err
+}
+
+// ServerGroupsAndResources returns the groups and supported resources for all groups and versions.
+func (d *memCacheClient) ServerGroupsAndResources() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	return discovery.ServerGroupsAndResources(d)
+}
+
+func (d *memCacheClient) ServerGroups() (*metav1.APIGroupList, error) {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:122]: Add comment
```golang
// ServerGroups returns the supported groups, with information like supported versions and the
// preferred version.
```

File: internal/kubernetes/cached/memory/memcache.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+func (d *memCacheClient) Reset() {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	d.fresh = true
+	d.invalidated = false
+	d.ourEntries = map[string]struct{}{}
+}
+
+func NewCachedDiscoveryClientForConfig(config *restclient.Config, ttl time.Duration) (CachedDiscoveryClient, error) {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:232]: Add comment
```golang
// NewCachedDiscoveryClientForConfig creates a new DiscoveryClient for the given config, and wraps
// the created client in a CachedDiscoveryClient. The provided configuration is updated with a
// custom transport that understands cache responses.
```

File: internal/kubernetes/cached/memory/memcache.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return newCachedDiscoveryClient(discoveryClient, ttl), nil
+}
+
+func NewCachedDiscoveryClient(delegate discovery.DiscoveryInterface, ttl time.Duration) CachedDiscoveryClient {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:248]: Add comment
```golang
// NewCachedDiscoveryClient creates a new CachedDiscoveryClient which caches
 // discovery information in memory and will stay up-to-date if Invalidate is
 // called with regularity.
 ```

File: internal/kubernetes/cached/memory/memcache.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	return newCachedDiscoveryClient(delegate, ttl)
+}
+
+// NewCacheClient creates a new CachedDiscoveryInterface which caches
+// discovery information in memory and will stay up-to-date if Invalidate is
+// called with regularity.
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:254]: Remove

File: internal/kubernetes/controller.go [hunk 5]
```diff
@@ -28,11 +35,28 @@ func NewController() Controller {
...
 
+// UseDiskCache sets the controller to generate clients that use
+// disk cache instead of memory cache for discovery and HTTP responses.
+func UseDiskCache() {
+	useDiskCache = true
+}
+
 // NewClient returns a new dynamic Kubernetes client with a default
 // disk cache directory of /var/kube/cache. This is where the client
 // stores and references its discovery of the Kubernetes API server.
```
Comment (dmrogers7) [inline internal/kubernetes/controller.go:46]: Not accurate any more

File: internal/kubernetes/controller.go [hunk 6]
```diff
@@ -51,9 +75,99 @@ const (
 	// Default cache directory.
 	cacheDir       = "/var/kube/cache"
 	defaultTimeout = 180 * time.Second
-	ttl            = 10 * time.Minute
+	ttl            = 2 * time.Minute
```
Comment (dmrogers7) [inline internal/kubernetes/controller.go:78]: Why was this shortened?

File: internal/kubernetes/cached/memory/memcache.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+func (d *memCacheClient) ServerPreferredNamespacedResources() ([]*metav1.APIResourceList, error) {
+	return discovery.ServerPreferredNamespacedResources(d)
+}
+
+func (d *memCacheClient) ServerVersion() (*version.Info, error) {
+	return d.delegate.ServerVersion()
+}
+
+func (d *memCacheClient) OpenAPISchema() (*openapi_v2.Document, error) {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:200]: Add comment
```golang
// OpenAPISchema retrieves and parses the swagger API schema the server supports.
```

File: internal/kubernetes/cached/memory/memcache.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+	return d.delegate.OpenAPISchema()
+}
+
+func (d *memCacheClient) Fresh() bool {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:204]: Add comment
```golang
// Fresh is supposed to tell the caller whether or not to retry if the cache
// fails to find something (false = retry, true = no need to retry).
```

File: internal/kubernetes/cached/memory/memcache.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+func (d *memCacheClient) ServerPreferredResources() ([]*metav1.APIResourceList, error) {
+	return discovery.ServerPreferredResources(d)
+}
+
+func (d *memCacheClient) ServerPreferredNamespacedResources() ([]*metav1.APIResourceList, error) {
+	return discovery.ServerPreferredNamespacedResources(d)
+}
+
+func (d *memCacheClient) ServerVersion() (*version.Info, error) {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:196]: Add comment
```golang
// ServerVersion retrieves and parses the server's version (git version).
```

File: internal/kubernetes/cached/memory/memcache.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+	return cachedEntry, nil
+}
+
+func (d *memCacheClient) createCachedEntry(key string, entry interface{}) {
+	d.entries[key] = entry
+	d.ourEntries[key] = struct{}{}
+	d.expirations[key] = time.Now().Add(d.ttl)
+}
+
+func (d *memCacheClient) RESTClient() restclient.Interface {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:184]: Add comment
```golang
// RESTClient returns a RESTClient that is used to communicate with API server
// by this client implementation.
```

File: internal/kubernetes/cached/memory/memcache.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+	return d.delegate.RESTClient()
+}
+
+func (d *memCacheClient) ServerPreferredResources() ([]*metav1.APIResourceList, error) {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:188]: Add comment
```golang
// ServerPreferredResources returns the supported resources with the version preferred by the
// server.
```

File: internal/kubernetes/cached/memory/memcache.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+	return discovery.ServerPreferredResources(d)
+}
+
+func (d *memCacheClient) ServerPreferredNamespacedResources() ([]*metav1.APIResourceList, error) {
```
Comment (dmrogers7) [inline internal/kubernetes/cached/memory/memcache.go:192]: Add comment
```golang
// ServerPreferredNamespacedResources returns the supported namespaced resources with the
// version preferred by the server.
```

File: internal/kubernetes/controller.go (hunk 6 above)
Comment (billiford) [reply internal/kubernetes/controller.go:78]: oops! Testing purposes.

**Review by dmrogers7 — APPROVED** [review]

**Comment by Bryce-huang** [comment]
> I'm still having an issue where new CRDs are not triggering API discovery. image

same issue in latest version

**Pull request merged**